src/authorization/access_control.rs

```diff
@@ -0,0 +1,209 @@
+    /// `None` means that `User::Anyone` permissions apply.
+    pub fn is_allowed(&self, access: AccessType) -> Option<bool> {
+        match access {
+            AccessType::Read => Some(true), // It's Public data, so it's always allowed to read it.
+            _ => match self.status.get(&access) {
+                Some(true) => Some(true),
+                Some(false) => Some(false),
+                None => None,
+            },
+        }
+    }
+}
+
+pub trait AccessListTrait: Clone + Eq + Ord + Hash + Serialize + DeserializeOwned {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool;
+    fn expected_data_version(&self) -> u64;
+    fn expected_owners_version(&self) -> u64;
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PrivateAccessList {
+    pub access_list: BTreeMap<PublicKey, PrivateUserAccess>,
+    /// The expected index of the data at the time this grant status change is to become valid.
+    pub expected_data_version: u64,
+    /// The expected index of the owners at the time this grant status is to become valid.
+    pub expected_owners_version: u64,
+}
+
+impl PrivateAccessList {
+    pub fn access_list(&self) -> &BTreeMap<PublicKey, PrivateUserAccess> {
+        &self.access_list
+    }
+}
+
+impl AccessListTrait for PrivateAccessList {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
+        match self.access_list.get(user) {
+            Some(access_status) => access_status.clone().is_allowed(access),
+            None => false,
+        }
+    }
+
+    fn expected_data_version(&self) -> u64 {
+        self.expected_data_version
+    }
+
+    fn expected_owners_version(&self) -> u64 {
+        self.expected_owners_version
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, PartialOrd, Ord, Eq, Hash, Debug)]
+pub struct PublicAccessList {
+    pub access_list: BTreeMap<User, PublicUserAccess>,
+    /// The expected index of the data at the time this grant status change is to become valid.
+    pub expected_data_version: u64,
+    /// The expected index of the owners at the time this grant status change is to become valid.
+    pub expected_owners_version: u64,
+}
+
+impl PublicAccessList {
+    fn is_allowed_(&self, user: &User, access: AccessType) -> Option<bool> {
+        match self.access_list.get(user) {
+            Some(status) => match status.clone().is_allowed(access) {
+                Some(true) => Some(true),
+                Some(false) => Some(false),
+                None => None,
+            },
+            _ => None,
+        }
+    }
+
+    pub fn access_list(&self) -> &BTreeMap<User, PublicUserAccess> {
+        &self.access_list
+    }
+}
+
+impl AccessListTrait for PublicAccessList {
+    fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
+        match self.is_allowed_(&User::Specific(*user), access) {
```

Comment:
Probably more idomatic and clearer:

```rust
     fn is_allowed(&self, user: &PublicKey, access: AccessType) -> bool {
        self.is_allowed_(&User::Specific(*user), access)
            .or_else(|| self.is_allowed_(&User::Anyone, access))
            .unwrap_or(false)
    } 
```

Reply:
This is exactly the kind of comments I was hoping for. Great!

Reply:
Aah, bad me..! I didn't check the logics or run the tests before committing this change :). 

It actually does not do the same thing.